libsolidity/formal/BMC.cpp

```diff
@@ -290,109 +291,216 @@ bool BMC::visit(Conditional const& _op)
 	return false;
 }
 
-// Here we consider the execution of two branches:
-// Branch 1 assumes the loop condition to be true and executes the loop once,
-// after resetting touched variables.
-// Branch 2 assumes the loop condition to be false and skips the loop after
-// visiting the condition (it might contain side-effects, they need to be considered)
-// and does not erase knowledge.
-// If the loop is a do-while, condition side-effects are lost since the body,
-// executed once before the condition, might reassign variables.
-// Variables touched by the loop are merged with Branch 2.
+// Unrolls while or do-while loop
 bool BMC::visit(WhileStatement const& _node)
 {
-	auto indicesBeforeLoop = copyVariableIndices();
+	auto indicesBefore = copyVariableIndices();
+
+	m_context.pushSolver();
+	// variables touched by loop might change their value
+	// assume their values are not constant
 	m_context.resetVariables(touchedVariables(_node));
-	decltype(indicesBeforeLoop) indicesAfterLoop;
+
+	_node.condition().accept(*this);
+
+	// do not apply condition target on expressions
+	// that value might be changed by other loops
+	if (isRootFunction() && !isInsideLoop())
+		addVerificationTarget(
+			VerificationTargetType::ConstantCondition,
+			expr(_node.condition()),
+			&_node.condition()
+		);
+	m_context.popSolver();
+
+	resetVariableIndices(indicesBefore);
+
+	unsigned int bmcLoopIterations = m_settings.bmcLoopIterations.value_or(1);
+	smtutil::Expression broke(false);
+	smtutil::Expression loopCondition(true);
 	if (_node.isDoWhile())
 	{
-		indicesAfterLoop = visitBranch(&_node.body()).first;
-		// TODO the assertions generated in the body should still be active in the condition
-		_node.condition().accept(*this);
-		if (isRootFunction())
-			addVerificationTarget(
-				VerificationTargetType::ConstantCondition,
-				expr(_node.condition()),
-				&_node.condition()
+		for (unsigned int i = 0; i < bmcLoopIterations; ++i)
+		{
+			loopScopes.emplace();
+
+			auto indicesBefore = copyVariableIndices();
+			pushPathCondition(loopCondition);
+			_node.body().accept(*this);
```

Comment:
Shouldn't we run an iteration of the body before the condition in the `do-while` case?

Reply:
The code does that. Body goes first, then we accept the condition.